analyzer/protocol/ipsec/ipsec.spicy

```diff
@@ -0,0 +1,632 @@
+	rec: IPSecIKE();
+};
+
+type SharedInfo = unit {
+	var next_payload: PayloadType;
+};
+
+public type IPSecIKE = unit {
+	# Use this to track state
+	var shared_info: SharedInfo;
+
+	initiator_spi: bytes &size=8;
+	responder_spi: bytes &size=8;
+	next_payload: uint8 &convert=PayloadType($$) { self.shared_info.next_payload = $$; }
+	version: bitfield(8) {
+		maj: 4..7;
+		min: 0..3;
+	};
+	exchange_type: uint8 &convert=ExchangeType($$);
+	flags: bitfield(8) {
+		E: 0;
+		C: 1;
+		A: 2;
+		I: 3;
+		V: 4;
+		R: 5;
+		X2: 6..7;
+	};
+	message_id: uint32;
+	length: uint32;
+	payloads_v1: IPSecIKEv1_Payload(self.shared_info, self.message_id)[] &size=self.length-28 if (self.version.maj == 1 && !self.flags.E);
+	payloads_v2: IPSecIKEv2_Payload(self.shared_info, self.message_id)[] &size=self.length-28 if (self.version.maj == 2);
+} &requires=( (self.version.maj == 1 || self.version.maj == 2) && self.version.min == 0);
+
+type IPSecIKEv2_Payload = unit(inout shared_info: SharedInfo, message_id: uint32) {
+	var message_id: uint32 = message_id;
+
+	var this_payload: PayloadType = shared_info.next_payload;
+
+	header: IPSecIKE_Payload_Header { shared_info.next_payload = $$.next_payload; }
+
+	switch ( self.this_payload ) {
+		PayloadType::SA -> payload_sa: IPSecIKEv2_SA_Payload(message_id) &size=self.header.payload_length-4;
+		PayloadType::KE -> payload_ke: IPSecIKEv2_KE_Payload(message_id) &size=self.header.payload_length-4;
+		PayloadType::IDi -> payload_idi: IPSecIKEv2_IDi_Payload(message_id) &size=self.header.payload_length-4;
+		PayloadType::IDr -> payload_idr: IPSecIKEv2_IDr_Payload(message_id) &size=self.header.payload_length-4;
+		PayloadType::CERT -> payload_cert: IPSecIKEv2_CERT_Payload(message_id) &size=self.header.payload_length-4;
+		PayloadType::CERTREQ -> payload_certreq: IPSecIKEv2_CERTREQ_Payload(message_id) &size=self.header.payload_length-4;
+		PayloadType::AUTH -> payload_auth: IPSecIKEv2_AUTH_Payload(message_id) &size=self.header.payload_length-4;
+		PayloadType::Ni -> payload_ni: IPSecIKEv2_Ni_Payload(message_id) &size=self.header.payload_length-4;
+		PayloadType::N -> payload_n: IPSecIKEv2_N_Payload(message_id) &size=self.header.payload_length-4;
+		PayloadType::D -> payload_d: IPSecIKEv2_D_Payload(message_id) &size=self.header.payload_length-4;
+		PayloadType::V -> payload_v: IPSecIKEv2_V_Payload(message_id) &size=self.header.payload_length-4;
+		PayloadType::TSi -> payload_tsi: IPSecIKEv2_TSi_Payload(message_id) &size=self.header.payload_length-4;
+		PayloadType::TSr -> payload_tsr: IPSecIKEv2_TSr_Payload(message_id) &size=self.header.payload_length-4;
+		PayloadType::E -> payload_e: IPSecIKEv2_E_Payload(message_id) &size=self.header.payload_length-4;
+		PayloadType::CP -> payload_cp: IPSecIKEv2_CP_Payload(message_id) &size=self.header.payload_length-4;
+		PayloadType::EAP -> payload_eap: IPSecIKEv2_EAP_Payload(message_id) &size=self.header.payload_length-4;
+		PayloadType::NO_NEXT_PAYLOAD -> : bytes &size=0;
+		* -> : bytes &eod;
```

Comment:
assuming that this list of types is exhaustive - it would be kind of neat (but just neat, not really necessary) if one could throw a weird in Zeke if one encounters a different payload type.

Reply:
This could probably be done in more than just this spot.  Admittedly, I haven't used weird enough in anything I've written yet.